File: algorithms/sort/quick.py
#!/usr/bin/python

def partition(a, left, right):
    while left < right:
        if a[left] > a[right]:
            a[left], a[right] = a[right], a[left]
            right -= 1
        left += 1
    return a, right

def sort(a, left, right):
    a, mid = partition(a, left, right)
    if left < mid - 1:
        a = sort(a, left, mid - 1)
    if mid < right:
        a = sort(a, mid, right)
    return a

def test(l):
    print 'before:', l
    l = sort(l, 0, len(l) - 1)
    print 'after: ', l

def main():
    test([1])
    print '--'
    test([2, 1])
    print '--'
    test([2, 1, 3])
    print '--'
    test([9, 3, 2, 3, 99, 2])
    print '--'
    test([9, 9, 9, 9, 3])
    print '--'
    test([1, 2, 3, 4, 5])
    print '--'
    test([5, 4, 3, 2, 1])
    print '--'
    test([6, 5, 4, 3, 2, 1])

main()

File: algorithms/sort/quick.c
#include <stdio.h>
#include <stdlib.h>

void
partition(int *a, int left, int right, int *pivot)
{
	while (left < right) {
		if (a[left] > a[right]) {
			int tmp = a[left];
			a[left] = a[right];
			a[right] = tmp;
			--right;
		}
		++left;
	}
	*pivot = right;
}

void
sort(int *a, int left, int right)
{
	int pivot;
	partition(a, left, right, &pivot);
	if (left < pivot - 1)
		sort(a, left, pivot - 1);
	if (pivot < right)
		sort(a, pivot, right);
}

void
print_array(int *a, int n)
{
	while (n--) {
		printf("%d ", *a++);
	}
	putchar('\n');
	fflush(stdout);
}

int
main(int argc, char **argv)
{
	(void)argc;
	(void)argv;

	int a1[] = { 5, 2, 9 };
	int a2[] = { 8, 9 };
	int a3[] = { 1, 2, 3 };
	int a4[] = { 9, 9, 10, 4, 22, 23, 21 };
	int a5[] = { 1, 5, 3, 8, 9 };

	#define TEST(a) \
		printf(#a " before: "); \
		print_array(a, sizeof(a)/sizeof(*a)); \
		sort(a, 0, sizeof(a)/sizeof(*a) - 1); \
		printf(#a " after: "); \
		print_array(a, sizeof(a)/sizeof(*a));

	TEST(a1);
	TEST(a2);
	TEST(a3);
	TEST(a4);
	TEST(a5);

	return 0;
}
